File: src/satellite_control/cpp/orbital_dynamics.cpp
#include "orbital_dynamics.hpp"
#include <cmath>

namespace satellite_control {

// ============================================================================
// Clohessy-Wiltshire (CW) Dynamics
// ============================================================================

CWDynamics::CWDynamics(double mean_motion) : n_(mean_motion), n_sq_(mean_motion * mean_motion) {}

Eigen::Vector3d CWDynamics::compute_acceleration(const Eigen::Vector3d& position, const Eigen::Vector3d& velocity) const {
    double x = position.x();
    double z = position.z();
    double vx = velocity.x();
    double vy = velocity.y();
    
    // CW equations of motion (Hill's equations):
    // ax = 3n^2x + 2nvy  (Radial)
    // ay = -2nvx         (Along-track)
    // az = -n^2z         (Cross-track / Normal)
    
    double ax = 3.0 * n_sq_ * x + 2.0 * n_ * vy;
    double ay = -2.0 * n_ * vx;
    double az = -n_sq_ * z;

    return Eigen::Vector3d(ax, ay, az);
}

std::pair<Eigen::Matrix<double, 6, 6>, Eigen::Matrix<double, 6, 3>> CWDynamics::get_state_matrices(double dt) const {
    /*
     * Continuous-time A matrix (6x6)
     * State: [x, y, z, vx, vy, vz]
     * 
     * x_dot = vx
     * y_dot = vy
     * z_dot = vz
     * vx_dot = 3n^2x + 2nvy
     * vy_dot = -2nvx
     * vz_dot = -n^2z
     */
    Eigen::Matrix<double, 6, 6> A_cont = Eigen::Matrix<double, 6, 6>::Zero();
    
    // Identity blocks for position derivatives (dr/dt = v)
    A_cont(0, 3) = 1.0;
    A_cont(1, 4) = 1.0;
    A_cont(2, 5) = 1.0;

    // Dynamics blocks (dv/dt = f(r,v))
    A_cont(3, 0) = 3.0 * n_sq_;   // dvx/dx
    A_cont(3, 4) = 2.0 * n_;      // dvx/dvy
    A_cont(4, 3) = -2.0 * n_;     // dvy/dvx
    A_cont(5, 2) = -n_sq_;        // dvz/dz

    // B matrix (6x3) - input is acceleration [ax, ay, az]
    Eigen::Matrix<double, 6, 3> B_cont = Eigen::Matrix<double, 6, 3>::Zero();
    B_cont(3, 0) = 1.0;
    B_cont(4, 1) = 1.0;
    B_cont(5, 2) = 1.0;

    // Discretize using forward Euler: A_d = I + A*dt
    Eigen::Matrix<double, 6, 6> A_disc = Eigen::Matrix<double, 6, 6>::Identity() + A_cont * dt;
    Eigen::Matrix<double, 6, 3> B_disc = B_cont * dt;

    return {A_disc, B_disc};
}

std::pair<Eigen::Matrix<double, 16, 16>, Eigen::Matrix<double, 16, 9>> CWDynamics::get_mpc_dynamics_matrices(double dt) const {
    /*
     * Full 16-state dynamics, adding just the CW terms to an empty matrix.
     * The MPC controller will add this to the standard rigid body dynamics.
     * 
     * State indices in 16-state vector:
     * 0-2: pos (x,y,z)
     * 7-9: vel (vx,vy,vz)
     */
    Eigen::Matrix<double, 16, 16> A_cw = Eigen::Matrix<double, 16, 16>::Zero();

    // Position -> Velocity coupling (Gravity Gradient & Coriolis)
    
    // vx_dot += 3n^2x + 2nvy
    A_cw(7, 0) = 3.0 * n_sq_ * dt;  // dvx/dx
    A_cw(7, 8) = 2.0 * n_ * dt;     // dvx/dvy (Coriolis, note: vy corresponds to state index 8)

    // vy_dot += -2nvx
    A_cw(8, 7) = -2.0 * n_ * dt;    // dvy/dvx

    // vz_dot += -n^2z
    A_cw(9, 2) = -n_sq_ * dt;       // dvz/dz

    // B matrix is zero for CW gravity effects (it's passive dynamics)
    Eigen::Matrix<double, 16, 9> B_cw = Eigen::Matrix<double, 16, 9>::Zero();

    return {A_cw, B_cw};
}

// ============================================================================
// TwoBodyDynamics Implementation
// ============================================================================

TwoBodyDynamics::TwoBodyDynamics(double mu, double target_radius)
    : mu_(mu),
      target_radius_(target_radius),
      mean_motion_(std::sqrt(mu / (target_radius * target_radius * target_radius))),
      orbit_phase_(0.0),
      target_pos_(target_radius, 0.0, 0.0),
      target_vel_(0.0, mean_motion_ * target_radius, 0.0)
{
}

Eigen::Vector3d TwoBodyDynamics::compute_acceleration(const Eigen::Vector3d& abs_position) const {
    // Newton's law of gravitation: a = -μ/|r|³ × r
    double r_mag = abs_position.norm();
    if (r_mag < 1.0) {
        // Prevent singularity at center
        return Eigen::Vector3d::Zero();
    }
    double r_cubed = r_mag * r_mag * r_mag;
    return (-mu_ / r_cubed) * abs_position;
}

Eigen::Vector3d TwoBodyDynamics::compute_relative_acceleration(
    const Eigen::Vector3d& rel_position,
    const Eigen::Vector3d& rel_velocity,
    const Eigen::Vector3d& target_pos
) const {
    // Inspector absolute position = target + relative
    Eigen::Vector3d inspector_pos = target_pos + rel_position;
    
    // Gravitational accelerations
    Eigen::Vector3d a_inspector = compute_acceleration(inspector_pos);
    Eigen::Vector3d a_target = compute_acceleration(target_pos);
    
    // Relative acceleration (differential gravity)
    // Note: rel_velocity is unused here but kept for interface consistency
    (void)rel_velocity;
    
    return a_inspector - a_target;
}

void TwoBodyDynamics::propagate_target(double dt) {
    // Simple circular orbit propagation
    // For circular orbit: phase increases at rate n
    orbit_phase_ += mean_motion_ * dt;
    
    // Wrap phase to [0, 2π)
    while (orbit_phase_ >= 2.0 * M_PI) {
        orbit_phase_ -= 2.0 * M_PI;
    }
    
    // Update target position (circular orbit in x-y plane)
    double r = target_radius_;
    target_pos_.x() = r * std::cos(orbit_phase_);
    target_pos_.y() = r * std::sin(orbit_phase_);
    target_pos_.z() = 0.0;
    
    // Update target velocity (tangent to orbit)
    double v = mean_motion_ * r;
    target_vel_.x() = -v * std::sin(orbit_phase_);
    target_vel_.y() =  v * std::cos(orbit_phase_);
    target_vel_.z() = 0.0;
}

} // namespace satellite_control

File: src/satellite_control/physics/orbital_dynamics.py
"""
Orbital Dynamics Module

Implements Hill-Clohessy-Wiltshire (CW) equations for relative motion
around a target satellite in circular orbit.

CW Equations (relative motion in Hill/LVLH frame):
    ẍ = 2nẏ + 3n²x + ax    (Radial - away from Earth)
    ÿ = -2nẋ + ay          (Along-track - velocity direction)
    z̈ = -n²z + az          (Cross-track - out of plane)

where n = √(μ/a³) is the orbital mean motion and a is thrust acceleration.
"""

from dataclasses import dataclass
from typing import Optional, Tuple

import numpy as np

from src.satellite_control.config.orbital_config import OrbitalConfig


@dataclass
class CWDynamics:
    """
    Hill-Clohessy-Wiltshire relative motion dynamics.

    Computes accelerations due to gravity gradient effects
    for a chaser satellite relative to a target in circular orbit.

    V2.0.0: Uses C++ backend (`_cpp_physics`) for performance if available.
    """

    orbital_config: OrbitalConfig

    def __post_init__(self):
        """Initialize C++ backend or fallback to Python."""
        self.n = self.orbital_config.mean_motion
        self.n_sq = self.n**2
        self._cpp_backend = None

        try:
            from satellite_control.cpp._cpp_physics import CWDynamics as CppCWDynamics

            self._cpp_backend = CppCWDynamics(self.n)
        except ImportError:
            pass  # Fallback to pure Python (below)

    @staticmethod
    def calculate_acceleration(
        position: np.ndarray, velocity: np.ndarray, n: float
    ) -> np.ndarray:
        """
        Compute CW gravitational acceleration (stateless).

        Args:
            position: Relative position [x, y, z] [m]
            velocity: Relative velocity [vx, vy, vz] [m/s]
            n: Mean motion [rad/s]

        Returns:
            Acceleration [ax, ay, az] [m/s²]
        """
        # Unpack components
        x, _, z = position[0], position[1], position[2]
        vx, vy, _ = velocity[0], velocity[1], velocity[2]

        n_sq = n * n

        # CW Equations
        # ẍ = 2nẏ + 3n²x
        # ÿ = -2nẋ
        # z̈ = -n²z
        ax = 3 * n_sq * x + 2 * n * vy
        ay = -2 * n * vx
        az = -n_sq * z

        return np.array([ax, ay, az])

    def compute_acceleration(
        self,
        position: np.ndarray,
        velocity: np.ndarray,
    ) -> np.ndarray:
        """
        Compute CW gravitational acceleration.

        Args:
            position: Relative position [x, y, z] in Hill frame [m]
            velocity: Relative velocity [vx, vy, vz] in Hill frame [m/s]

        Returns:
            Gravitational acceleration [ax, ay, az] [m/s²]
        """
        if self._cpp_backend:
            return self._cpp_backend.compute_acceleration(position, velocity)

        return self.calculate_acceleration(position, velocity, self.n)

    def get_state_matrices(self, dt: float) -> Tuple[np.ndarray, np.ndarray]:
        """
        Get discrete-time state transition matrices for CW dynamics.

        State: [x, y, z, vx, vy, vz]

        Args:
            dt: Time step [s]

        Returns:
            (A, B) discrete-time matrices where x_{k+1} = A @ x_k + B @ u_k
        """
        if self._cpp_backend:
            return self._cpp_backend.get_state_matrices(dt)

        n = self.n
        n_sq = self.n_sq

        # Continuous-time A matrix
        # State: [x, y, z, vx, vy, vz]
        # Rows 0-2: velocity integration
        # Rows 3-5: acceleration equations
        A_cont = np.array(
            [
                [0, 0, 0, 1, 0, 0],
                [0, 0, 0, 0, 1, 0],
                [0, 0, 0, 0, 0, 1],
                [3 * n_sq, 0, 0, 0, 2 * n, 0],
                [0, 0, 0, -2 * n, 0, 0],
                [0, 0, -n_sq, 0, 0, 0],
            ]
        )

        # B matrix (input is acceleration)
        B_cont = np.zeros((6, 3))
        B_cont[3:, :] = np.eye(3)

        # Discretize (Forward Euler)
        # Note: For higher precision, use expm(A*dt), but Euler is sufficient for small dt
        A_disc = np.eye(6) + A_cont * dt
        B_disc = B_cont * dt

        return A_disc, B_disc

    def get_mpc_dynamics_matrices(self, dt: float) -> Tuple[np.ndarray, np.ndarray]:
        """
        Get 16-state dynamics matrices including CW terms for MPC.

        Integrates CW accelerations into the full satellite state.

        Args:
            dt: Time step [s]

        Returns:
            (A_cw, B_cw) matrices to add to existing MPC dynamics
        """
        if self._cpp_backend:
            return self._cpp_backend.get_mpc_dynamics_matrices(dt)

        n = self.n
        n_sq = self.n_sq

        # Additional A terms for velocity due to CW (affects positions 7-9)
        # State indices: 0-2(pos), 3-6(quat), 7-9(vel), ...
        A_cw = np.zeros((16, 16))

        # Position -> Velocity coupling
        # dvx/dx = 3n²
        A_cw[7, 0] = 3 * n_sq * dt
        # dvx/dvy = 2n (coriolis)
        A_cw[7, 8] = 2 * n * dt
        # dvy/dvx = -2n (coriolis)
        A_cw[8, 7] = -2 * n * dt
        # dvz/dz = -n²
        A_cw[9, 2] = -n_sq * dt

        return A_cw, np.zeros((16, 9))


def compute_cw_acceleration(
    position: np.ndarray,
    velocity: np.ndarray,
    mean_motion: float,
) -> np.ndarray:
    """
    Compute CW gravitational acceleration (standalone function).

    Delegates to CWDynamics logic.
    """
    return CWDynamics.calculate_acceleration(position, velocity, mean_motion)


def compute_cw_force(
    position: np.ndarray,
    velocity: np.ndarray,
    mass: float,
    orbital_config: Optional[OrbitalConfig] = None,
) -> np.ndarray:
    """
    Compute CW gravitational force for MuJoCo xfrc_applied.

    Args:
        position: Relative position [x, y, z] [m]
        velocity: Relative velocity [vx, vy, vz] [m/s]
        mass: Satellite mass [kg]
        orbital_config: Orbital configuration (default: LEO 400km)

    Returns:
        Force [Fx, Fy, Fz] [N]
    """
    n = orbital_config.mean_motion if orbital_config else OrbitalConfig().mean_motion
    accel = compute_cw_acceleration(position, velocity, n)
    return mass * accel
